R/buttons.R

```diff
@@ -0,0 +1,106 @@
+    class = c("button_spec", "list")
+  )
+}
+
+apply_button_spec <- function(status, value, defaults, ...) {
+  if (is.character(value) || is.null(value)) {
+    value <- list(label = value)
+  }
+  resolved <- defaults
+  resolved[names(value)] <- value
+
+  tagList(
+    tags$span(slot=paste0(status, "-icon"),
+      resolved[["icon"]]
+    ),
+    tags$span(slot=paste0(status, "-label"),
+      resolved[["label"]]
+    )
+  )
+}
+
+#' Button for launching longer-running operations
+#'
+#' `input_task_button` is a button whose value is initially zero, and increments
+#' by one each time it is pressed. It is similar to [shiny::actionButton()],
+#' except that upon click, it puts itself into a busy state that prevents the
+#' user from clicking it again. `update_task_button(id, busy = FALSE)` must be
+#' called from the server function to re-enable the button.
+#'
+#' @param id The `input` slot that will be used to access the value.
+#' @param label The label of the button while it is in ready (clickable) state;
+#'   usually a string.
+#' @param icon An optional icon to display next to the label while the button is
+#'   in ready state. See [fontawesome::fa_i()].
+#' @param busy_label The label of the button while it is busy.
+#' @param busy_icon The icon to display while the button is busy. By default, a
+#'   spinning "chasing arrows" icon is used.
+#' @param type One of the Bootstrap theme colors (`"primary"`, `"default"`,
+#'   `"secondary"`, `"success"`, `"danger"`, `"warning"`, `"info"`, `"light"`,
+#'   `"dark"`), or `NULL` to leave off the Bootstrap-specific button CSS classes
+#'   altogether.
+#' @param ... Named arguments become attributes to include on the `<button>`
+#'   element.
+#'
+#' @section Server value:
+#' An integer of class `"shinyActionButtonValue"`. This class differs from
+#' ordinary integers in that a value of 0 is considered "falsy".
+#' This implies two things:
+#'   * Event handlers (e.g., [shiny::observeEvent()], [shiny::eventReactive()]) won't execute on initial load.
+#'   * Input validation (e.g., [shiny::req()], [shiny::need()]) will fail on initial load.
+#'
+#' @export
+input_task_button <- function(id, label, icon = NULL,
+  busy_label = "Processing...", busy_icon = icon_spinner(), type = "primary",
+  ...) {
+
+  if (missing(type)) {
+    type <- "primary"
+  }
+  btn_class <- if (!is.null(type)) {
+    paste0("btn btn-", type)
+  }
+
+  tags$button(
+    id = id,
+    class = btn_class,
+    class = "bslib-task-button",
+    type = "button",
+
+    component_dependencies(),
+
+    htmltools::tag("bslib-task-button-contents",
+      list(
+        span(slot = "ready",
+          icon, label
+        ),
+        span(slot = "busy",
+          busy_icon, busy_label
+        )
+      )
+    ),
+
+    ...
+  )
+}
+
+#' @rdname input_task_button
+#' @export
+icon_spinner <- function() {
+  fontawesome::fa_i("refresh", class = "fa-spin", "aria-hidden" = "true")
```

Comment:
Could replace this with the result of `bsicons::bs_icon("arrow-clockwise")`?

https://icons.getbootstrap.com/icons/arrow-clockwise/

Reply:
Oh, cool! Do we have a feature to make it spin like the fontawesome version does?

Reply:
Also might be worth considering using a [Bootstrap spinner](https://getbootstrap.com/docs/5.3/components/spinners/) by default

Reply:
I don't think Bootstrap Icons have a spin class that would akin to `.fa-spin`, though, but we could add an appropriate utility class

Reply:
I tried the Bootstrap spinner, it almost looks good in `sm` size but the stroke is a little too thick IMHO. I like that the fontawesome one (and presumably bootstrap one) scales nicely with font size.

Are we trying to move away from fontawesome and towards bootstrap icons in general?

Reply:
I think in general we're trending toward Bootstrap icons -- it's a lighter weight library and designed to pair nicely with Bootstrap -- but I'm all for using what looks best. We do generally try to inline SVG icons so that the fontawesome or bsicons dependencies aren't strong deps.

Reply:
Yea, I'm not fundamentally against using `{fontawesome}`, but it does feel weird to export a thin wrapper around `fa_i()`. Also, assuming we are going to allow for customization of the icon, it feels like we should have a framework-agnostic way to spin the icon

Reply:
I wrote the thin wrapper around `fa_i()` mostly because inlining the full `fontawesome::fa_i("refresh", class = "fa-spin", "aria-hidden" = "true")` as the default value of the `busy_icon` parameter really ruined the readability of the `input_task_button` function signature.

And also it seemed weird for such a specific implementation detail to make it into the function signature. Not wrong from first principles, just aesthetically rubbed me the wrong way.

Reply:
I like the idea of our own spinning implementation but if it's OK with you I'd like to leave it off of this PR. It feels like we could definitely add it afterwards without any back compat issues.

Reply:
Given the above, I think we should:

1. Not export `icon_spinner()` and instead use `icon_spinner()` when the busy icon is not provided by a user.
2. Later come back and both inline the icon svg and add a utility class to spin icons.

Reply:
After discussion: switch to rlang::missing_arg(), have the `icon_busy` documentation mention what the default is (meaning, the actual code).